### PR TITLE
fetch_last_run helper function

### PR DESCRIPTION
**Pull request by agamm**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
A basic POC for abstracting the last run flow.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/AzureActiveDirectory/Integrations/AzureADIdentityProtection/AzureADIdentityProtection.py [hunk 1]
```diff
@@ -311,6 +311,60 @@ def build_filter(last_fetch, params):
...
+          uses last time as a parameter and returns the new time.
+        params (Dict[str, str]): Demisto configuration object
+        time_format_func (Callable): A user defined function to convert date
+                            time into the intergration specified date format.
+        integration_name (str): Integration name as prefix for debug logs.
+      Returns:
+        None
+    '''
+
+    log_prefix = f'[{integration_name}]' if integration_name else '[Debug]'
```
Comment (tkatzir) [inline Packs/AzureActiveDirectory/Integrations/AzureADIdentityProtection/AzureADIdentityProtection.py:335]: I'd keep the [Debug] log level, and add the integration_name, for standardization.

File: Packs/AzureActiveDirectory/Integrations/AzureADIdentityProtection/AzureADIdentityProtection.py [hunk 2]
```diff
@@ -311,6 +311,60 @@ def build_filter(last_fetch, params):
...
+    initial_offset = dateparser.parse(date_string=initial_offset_string)
+                                      #date_formats=[DATE_FORMAT])
+
+    if not last_run.get('last_time_run', None):
+        demisto.debug(f'{log_prefix} first run, setting to initial offset: '
+                      f'{initial_offset_string}')
+        last_run['last_time_run'] = initial_offset
+
+    # If the user supplied a time conversion function, use it.
+    if time_format_func:
```
Comment (tkatzir) [inline Packs/AzureActiveDirectory/Integrations/AzureADIdentityProtection/AzureADIdentityProtection.py:351]: Rename this hook method to something like `fetch_time_formatter`, or `fetch_time_format_hook`, or similar. It's a very generic name as it is.

File: Packs/AzureActiveDirectory/Integrations/AzureADIdentityProtection/AzureADIdentityProtection.py [hunk 3]
```diff
@@ -311,6 +311,60 @@ def build_filter(last_fetch, params):
...
+    # Calling the user's fetch function
+    demisto.debug(f'{log_prefix} fetching...')
+    res = fetch_func(last_run)
+
+    new_last_run = {
+        'last_time_run': res.get('last_time_run', initial_offset),
+        'last_etag': res.get('last_etag',
+                                  last_run.get('last_etag', None))
+    }
+    demisto.debug(f'{log_prefix} done fetching! setting new last run: {new_last_run}')
```
Comment (tkatzir) [inline Packs/AzureActiveDirectory/Integrations/AzureADIdentityProtection/AzureADIdentityProtection.py:363]: Maybe log `last_etag` as well?

File: Packs/AzureActiveDirectory/Integrations/AzureADIdentityProtection/AzureADIdentityProtection.py (hunk 2 above)
Comment (tkatzir) [inline Packs/AzureActiveDirectory/Integrations/AzureADIdentityProtection/AzureADIdentityProtection.py:351]: Maybe create a trivial implementation of this hook method, and document it with docstring and sample output, so customizers would know how to develop it.

File: Packs/AzureActiveDirectory/Integrations/AzureADIdentityProtection/AzureADIdentityProtection.py [hunk 4]
```diff
@@ -329,32 +383,32 @@ def date_str_to_azure_format(date_str):
...
+        detections: list = risk_detection_list_raw.get('value', [])
 
-    incidents, latest_detection_time = detections_to_incidents(detections, last_fetch_datetime=last_fetch_datetime)
-    demisto.debug(f'[AzureIdentityProtection] Fetched {len(incidents)} incidents')
+        last_fetch_datetime: datetime = datetime.strptime(last_fetch, DATE_FORMAT)
+        incidents, latest_detection_time = detections_to_incidents(detections, last_fetch_datetime=last_fetch_datetime)
+        demisto.debug(f'[AzureIdentityProtection] Fetched {len(incidents)} incidents')
+        latest_detection_time = latest_detection_time.strftime(DATE_FORMAT)
+        return {
+            'last_time_run': latest_detection_time
```
Comment (tkatzir) [inline Packs/AzureActiveDirectory/Integrations/AzureADIdentityProtection/AzureADIdentityProtection.py:404]: Missing last_etag?

**Review by tkatzir — CHANGES_REQUESTED** [review]

File: Packs/AzureActiveDirectory/Integrations/AzureADIdentityProtection/AzureADIdentityProtection.py (hunk 3 above)
Comment (agamm) [reply Packs/AzureActiveDirectory/Integrations/AzureADIdentityProtection/AzureADIdentityProtection.py:363]: It already includes it in the `new_last_run` object.

File: Packs/AzureActiveDirectory/Integrations/AzureADIdentityProtection/AzureADIdentityProtection.py (hunk 4 above)
Comment (agamm) [reply Packs/AzureActiveDirectory/Integrations/AzureADIdentityProtection/AzureADIdentityProtection.py:404]: You don't use etag so no need to put it.

**Pull request closed without merging**